### PR TITLE
Improve meta handling and add lubimyczytac.pl meta provider

### PR DESCRIPTION
**Pull request by collerek**
Thank you for this great project!

I made some improvements over how metadata are fetched and added a new provider:

- Add interfaces for book info (`MetaRecord`) and source information (`MetaSourceInfo`) to unify the available fields, provide code autocompletion in IDE, and document for other developers what are the expected fields meta should provide
- Unify all meta providers to use those interfaces
- Fill also `series_index` from meta (now skipped)
- Fill also `languages` from meta (now skipped)
- Update/ add `identifiers` -> at least meta id, ISBN if available (now skipped)
- Add new meta provider (LubimyCzytac.pl) Close: #1948
- Extract more info from Google Books (`languages`, `ISBN`)
- Extract more info from Google Scholar (set `identifier` and `id` to publication URL)
- Clean and URL encode all search queries

Let me know what you think!

### CODE REVIEW

**Comment by collerek** [comment]
Hi @OzzieIsaacs, saw your comment in #2204, should I close this PR, as you want to change this totally into plugins?

lubimyczytać also needs a second call for every book but it's done in a `ThreadPool` so results are quite fast (few sec max).
Same is needed for comicvine (but I don't have comics so didn't "fix" this) as described in #1705 as the api results are partial only.

**Comment by OzzieIsaacs** [comment]
I checked it, this one was fast. You also did a lot of changes so I need time for checking. I didn‘t want the other pr to be closed. I need time to think on how to deal with metadata provider in future

**Comment by collerek** [comment]
@OzzieIsaacs any chance of getting this merged?

**Comment by OzzieIsaacs** [comment]
I plan a release this weekend, afterwards I will start on this one and on the cover thumbnails one.
This one is a bigger one and it's in an rather untested part of calibre-web, so I have to write also some new tests.

File: cps/services/Metadata.py [hunk 1]
```diff
@@ -15,13 +15,93 @@
...
+        self, query: str, generic_cover: str = "", locale: str = "en"
+    ) -> Optional[List[MetaRecord]]:
+        pass
+
+    @staticmethod
+    def get_title_tokens(
+        title: str, strip_joiners: bool = True
+    ) -> Generator[str, None, None]:
+        """
+        Taken from calibre source code
```
Comment (OzzieIsaacs) [inline cps/services/Metadata.py:74]: Could you please do me one last favor and give some reference (which file/line, which commit or similar) to find the code ( I don't want to search for it by myself) 
thanks

Comment (collerek) [reply cps/services/Metadata.py:74]: Sure thing.

It's a simplified (cut out what is unnecessary) version of 
https://github.com/kovidgoyal/calibre/blob/99d85b97918625d172227c8ffb7e0c71794966c0/src/calibre/ebooks/metadata/sources/base.py#L363-L367

(src/calibre/ebooks/metadata/sources/base.py - lines 363-398)

**Comment by collerek** [comment]
@OzzieIsaacs Thanks for merging and your hard work! :)
Do you have an estimate of when will be the next release with these features?

**Comment by OzzieIsaacs** [comment]
Next weekend (my windows pc crashed last weekend, I have to get it up running) maybe

**Pull request merged**